torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,171 @@
+        >>> train, valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0)
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+        >>> # You can also specify a target key if you only need a specific group of samples
+        >>> train = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='train')
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> # Be careful to use the same seed as before when specifying `target` to get the correct split.
+        >>> valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='valid')
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+    """
+
+    def __new__(
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
+        total_length: Optional[int] = None,
+        target: Optional[T] = None,
+    ):
+        if total_length is None:
+            try:
+                # TODO: This is an issue for DataPipes which only have runtime lengths. Revisit to see if this
+                #       is problematic.
+                total_length = len(source_datapipe)
+            except TypeError:
+                raise TypeError(
+                    "RandomSplitter needs `total_length`, but it is unable to infer it from "
+                    f"the `source_datapipe`: {source_datapipe}."
+                )
+        container = _RandomSplitterIterDataPipe(source_datapipe, total_length, weights, seed)
+        if target is None:
+            return [SplitterIterator(container, k) for k in list(weights.keys())]
+        else:
+            if target in weights.keys():
+                return SplitterIterator(container, target)
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length = total_length
+        self._seed = seed
+        self.norm_weights = self.normalize_weights(weights, total_length)
+        self.keys = list(weights.keys())
+        self.key_to_index = {k: i for i, k in enumerate(self.keys)}
+        self.weights = [self.norm_weights[k] for k in self.keys]
+        self._rng = random.Random(self._seed)
+
+    def draw(self):
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        self.weights[self.key_to_index[selected_key]] -= 1
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights, total_length: int):
+        total_weight = sum(weights.values())
+        return {k: round(float(w) * total_length / total_weight) for k, w in weights.items()}
```

Comment:
And, using `round` might lead to the sum of weights not equal to `total_length`.

Reply:
You are right. This also means we have to renormalize the weights after one of the keys run out. I have updated my implementation. Please verify.

Reply:
I am unsure. If one key runs out, why do we need to renormalize? If one weight goes to 0, the corresponding key should never be returned by `random.choices` IMHO.

Reply:
Let's say we have `total_length=10` and `weights=[3.33, 3.33, 3.33]`.
We draw 4 '0's in a row, the remaining weights will be `weights=[-0.67, 3.33, 3.33]`.
We can set the negative value to 0 and get `weights=[0, 3.33, 3.33]`. Now the remaining number of draws is 6 but the weights sum up to `6.66`. 

It should not impact the result/output, but if we choose to normalize, it will bring the sum back to `6` to match the remaining number of draws. I can take it out since it is technically unnecessary computation.

Reply:
I think I will take it out, because it will greatly simplify the other functions. For example, `normalize_weights` will no longer have to consider the `list` case.

Reply:
Then, how about we always convert `weights` to a list of integer. For example, we do `floor` over all the elements except the last one. And, the rest of elements send to the last key. Then, we don't need to normalize and the weight is kept as integer then no negative value will happen.

Reply:
Two things:
1. I was wrong about not needing to normalize after a weight hitting negative. Actually it is necessary.
    Example: Splitting 10 elements to 9 DPs: [1.11, 1.11, 1.11, 1.11, 1.11, 1.11, 1.11, 1.11, 1.11]
    Without normalization, you may end up with 2 or more DPs with more than one element, while the ideal case is 8 
    DPs with 1 element and 1 DP with 2.
2. I considered your suggestion to `floor` all but the last one. It changes the weights from its true distributino.
    Example: 10 elements split into two DPs with weights [0.89, 0.11], most of the time you want 9 element in the first 
    and 1 element second DP.

Because of these complications, I am going to add a note, telling users that if they want certainty, please provide integer weights that sum up to `total_length`.

Reply:
IIUC, the current implementation is similar to doing `ceiling` to all keys until the last requested. That's why we have to do re-normalization when each time one key is depleted. It's fine but I feel like we can do it in advance without doing normalization multiple times.